### DISCUSSION TITLE
Is there any chance for the hybrid cache to get phantom entry?

### DISCUSSION THREAD

**Discussion by MrCroxx** [facebook/CacheLib#284] [Q&A]
Hi, I'm new to CacheLib and using it to build my app. And I'm interested in reading CacheLib's source code. 

I found the `BlockCache` (large entry cache) of the hybrid cache engine `navy` doesn't persist tombstones when removing entries. Is there any chance for the hybrid cache get a phantom entry after recovery?

**Comment by haowu14** [comment]
What's accessible in BlockCache is only those entries in the BlockCache Index. So as long as an entry is removed from the index, it won't be accessible again.

The tombstone in NvmCache code is used for synchronization for inflight lookups/puts/deletes only. Before a proper shut down all inflight NVM operations will be drained, so it doesn't need to be persisted.

**Reply by MrCroxx** [reply]
But are indices persisted? I thought they are rebuilt from entries by reading through the regions.

**Reply by haowu14** [reply]
Yes. They were persisted in the cache directory. On startup, the recover() call read from the thrift file and got a list of buckets: https://github.com/facebook/CacheLib/blob/main/cachelib/navy/block_cache/Index.cpp#L186

**Reply by MrCroxx** [reply]
IIUC, the indices don't persist itself or log its mutatuon per kv. Is there a chance that after a deletion but before the new indices are persisted the server crashes and recovers? In this case, will a deleted entry comes back to life?

**Reply by haowu14** [reply]
I guess the term persisted is misleading. CacheLib is not a persisted storage in the sense that it won't recover from power loss or crashes.

If the process does not get proper shut down (didn't call CacheAllocator::[shutDown](https://github.com/facebook/CacheLib/blob/main/cachelib/allocator/CacheAllocator.h#L964)() , the cache will be dropped.

**Reply by MrCroxx** [reply]
Oh, I see. Then it makes sense.

**Reply by MrCroxx** [reply]
Thanks for your patience. Have a nice day. 🙏

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
